### ISSUE TITLE
Autocomplete not ignoring preprocessor directives in function header

### ISSUE THREAD

**Issue by monkey0506** [adventuregamestudio/ags#364]
If a function header includes a `#ifdef` or `#ifndef` block that is not preceded by an optional argument, then autocomplete includes the directive as part of the previous or current argument's definition:

```
int myFunc(
  #ifdef SCRIPT_API_v340
  int param=0
  #endif
  )
{
}
```

Will autocomplete as:

```
int myFunc(optional #ifdef SCRIPT_API_v340
int param)
```

And

```
int myFunc2(int a
  #ifdef SCRIPT_API_v340
  , int param=0
  #endif
  )
{
}
```

Will autocomplete as:

```
int myFunc2(int a
#ifdef SCRIPT_API_v340, optional int param)
```

This behavior does not happen if the preprocessor directive follows an optional argument (e.g., it doesn't happen to `Character.ChangeRoom`).

Additionally, it includes the ROL, so even autocomplete tags like `$AUTOCOMPLETEIGNORE$` are not utilized or seen as comments, but included as part of the function header.

**Comment by rofl0r** [comment]
just a thought... when seeing this: `#ifdef SCRIPT_API_v340` it appears to me that `#if SCRIPT_API >= 340` would be a way more powerful idiom...

**Comment by monkey0506** [comment]
@rofl0r you are correct, but the AGS preprocessor doesn't have an `#if` directive (currently).

**Comment by ivan-mogilko** [comment]
Speaking of autocomplete, I stumbled upon it when doing some changes to the script, and I found that it actually duplicates precompiler instead of reusing it.
I wrote a quick note in the code: https://github.com/adventuregamestudio/ags/blob/master/Editor/AGS.Editor/Panes/ScintillaWrapper.cs#L1742
Maybe if someone is going to work on this could take this into consideration.

**Comment by gurok** [comment]
Why was this closed? As far as I know, it's still an issue.

**Comment by ivan-mogilko** [comment]
The culprit for this bug is [`AutoComplete.ConstructCache()`](https://github.com/adventuregamestudio/ags/blob/ea4e6762236a13621ba0771b9d5d8ac4bcf33ee8/Editor/AGS.Editor/AutoComplete.cs#L142), which tries to handle preprocessor options in parallel with parsing script constructs, and thus potentially missing number of cases.

The solution could be to preprocess first, and then parse resulting script.

**Comment by ericoporto** [comment]
> The solution could be to preprocess first, and then parse resulting script

I think we can do that as long we keep the define macros, the triple comments, and replace any removed character with a blank space to keep the character index matching.

I think this may be done in the existing preprocessor code, as an option.

**Comment by ericoporto** [comment]
I tried to rewrite the header that ships with the editor to prevent this issue at least in our own script api

https://github.com/ericoporto/ags/commit/03b49495a637d3034b1ee35fb0f41277001d6a22

But it turns out the nesting inside else clauses is not working properly either. :/
